daemon/internal/newrelic/app.go

```diff
@@ -338,3 +340,75 @@ func (app *App) Inactive(threshold time.Duration) bool {
 	}
 	return time.Since(app.LastActivity) > threshold
 }
+
+// filter seen php packages data to avoid sending duplicates
+//
+// the `App` structure contains a map of PHP Packages the reporting
+// application has encountered.
+//
+// the map of packages should persist for the duration of the
+// current connection
+//
+// takes the `PhpPackages.data` byte array as input and unmarshals
+// into an anonymous interface array
+//
+// the JSON format received from the agent is:
+// 		[["package_name","version",{}],...]
+//
+// for each entry, assign the package name and version to the `PhpPackagesKey`
+// struct and use the key to verify data does not exist in the map. If the
+// key does not exist, add it to the map and the array of 'new' packages.
+//
+// convert the array of 'new' packages into a byte array representing
+// the expected data that should match input, minus the duplicates.
+func (app *App) filterPhpPackages(data []byte) []byte {
+	if data == nil {
+		return nil
```

Comment:
When this returns nil, are we still harvesting the package data and sending update_loaded_modules with empty package info?

Reply:
Currently [this code](https://github.com/newrelic/newrelic-php-agent/blob/main/daemon/internal/newrelic/php_packages.go#L65) checks if `packages`, `packages.data` or `packages.numSeen` are `nil` or `0` prior to mashalling the data for the endpoint/collector and results in returning the empty payload `["Jars",[]]`. I've attempted to preserve this behavior for this PR.

Reply:
So majority of time, we'll be sending an empty payload?
Since it's not required to be sent as part of the harvest. if the payload is empty, should we still be sending it?
Otherwise, we'll still sending (most likely) an empty payload of the optional update_loaded_modules with every harvest that the backend will still have to process.

Reply:
If `newrelic.vulnerability_management.package_detection.enabled` is set to `true`, we should be sending payloads until we've managed to instrument all the supported libraries/packages, then the rest of the payloads will be empty. This is also on a per-process and per-connect basis, so it should reset at least once every day. We could update logic to not send anything if nothing is detected, but I think that would be best done in a separate PR since it modifies core packages code outside of this feature.

Reply:
Having it another PR is totally fine, but important to get done as part of this effort.  Is this sent on fast harvest or regular harvest?  Once a significant amount of customers are on this agent version, it would be sending an egregious amount of empty payloads every day for the backend to process and discard.

Reply:
Should only be sent on regular harvest cycle of 1 minute. Once this PR is merged I can put up another one adding the feature to skip reporting to this endpoint on `nil` `data`.

Reply:
On second pass through the code and this may actually not be the case- it seems like the `PhpPackages` struct implements the `PayloadCreator` interface which calls the `Empty` method (described above) when determining whether or not to send the harvest in `considerHarvestPayload`, which if `Empty` is true (aka `data` is `nil`), `considerHarvestPayload` simply returns before calling `harvestPayload`. The "empty" payload of `["Jars", []]` may only be for internal audit (logging) purposes.

Reply:
Thanks for addition tests to confirm behavior!!

[test: add test for nil package data for harvest](https://github.com/newrelic/newrelic-php-agent/pull/871/commits/6cc5b187f12e80c5c5d39da75398514e774de542)
[6cc5b18](https://github.com/newrelic/newrelic-php-agent/pull/871/commits/6cc5b187f12e80c5c5d39da75398514e774de542)

[test: add harvest test for valid package data](https://github.com/newrelic/newrelic-php-agent/pull/871/commits/8ecb0fdbb8e45ede21eb4419c1f29c47362114e8)
[8ecb0fd](https://github.com/newrelic/newrelic-php-agent/pull/871/commits/8ecb0fdbb8e45ede21eb4419c1f29c47362114e8)